cmake/Dependencies.cmake

```diff
@@ -1169,6 +1169,40 @@ if(USE_GLOO)
   endif()
 endif()
 
+if(USE_TENSORPIPE)
+  if(MSVC)
+    message(WARNING "Tensorpipe cannot be used on Windows.")
+  else()
+    if(CAFFE2_LINK_LOCAL_PROTOBUF)
+      set(TP_EXTRA_CMAKE_ARGS -DProtobuf_DIR=${PROJECT_BINARY_DIR}/third_party/protobuf/cmake/${CMAKE_INSTALL_LIBDIR}/cmake/protobuf)
+    endif()
+
+    include(ExternalProject)
+
+    ExternalProject_Add(tensorpipe-build
+      PREFIX tensorpipe
+      SOURCE_DIR ${PROJECT_SOURCE_DIR}/third_party/tensorpipe
+      BUILD_BYPRODUCTS tensorpipe/${CMAKE_INSTALL_LIBDIR}/libuv_a.a tensorpipe/${CMAKE_INSTALL_LIBDIR}/libtensorpipe.a
```

Comment:
Question: (not requesting for change) this will build TensorPipe's own libuv? IIRC, Gloo would find libuv using `PkgConfig` at least on MacOS. Will these two libuv libs cause any problem?

Reply:
> Question: (not requesting for change) this will build TensorPipe's own libuv?
Yes, this is the case.
> IIRC, Gloo would find libuv using PkgConfig at least on MacOS.
This is supported by TensorPipe as well, but requires `pkg-config` to be available.
> Will these two libuv libs cause any problem?
Probably not, but we might as well ensure we use the same one. Two strategies here: either have libuv pre-installed as a requirement for building pytorch, or have pytorch's build explicitly depend on libuv (when `USE_DISTRIBUTED=ON`) as a submodule.

Reply:
At the moment, however, if libuv is installed on the system and can be found through pkg-config, both Gloo and TensorPipe will use that version, right? Could we say that this is enough for this initial version?

Reply:
@lw At the moment we force building a custom libuv within tensorpipe when building pytorch, but this can be changed.